// dcorejs/lib/model/proposal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ProposalError;
(function (ProposalError) {
    ProposalError["database_operation_failed"] = "database_operation_failed";
    ProposalError["chain_operation_failed"] = "chain_operation_failed";
    ProposalError["transaction_broadcast_failed"] = "transaction_broadcast_failed";
    ProposalError["account_does_not_exist"] = "account_does_not_exist";
    ProposalError["account_fetch_failed"] = "account_fetch_failed";
    ProposalError["transfer_sender_account_does_not_exist"] = "transfer_sender_account_does_not_exist";
    ProposalError["transfer_receiver_account_does_not_exist"] = "transfer_receiver_account_does_not_exist";
    ProposalError["asset_not_found"] = "asset_not_found";
    ProposalError["propose_object_not_found"] = "propose_object_not_found";
    ProposalError["connection_failed"] = "connection_failed";
    ProposalError["syntactic_error"] = "syntactic_error";
})(ProposalError = exports.ProposalError || (exports.ProposalError = {}));
